src/v/coproc/pacemaker.cc

```diff
@@ -0,0 +1,196 @@
+        return errc::invalid_ingestion_policy;
+    }
+    if (model::is_materialized_topic(topic)) {
+        return errc::materialized_topic;
+    }
+    if (model::validate_kafka_topic_name(topic).value() != 0) {
+        return errc::invalid_topic;
+    }
+    return errc::success;
+}
+
+ss::future<> pacemaker::do_add_source(
+  script_id id,
+  ntp_context_cache& ctxs,
+  std::vector<errc>& acks,
+  const std::vector<std::pair<model::topic, topic_ingestion_policy>>& tnss) {
+    return ss::do_for_each(
+      tnss,
+      [this, &ctxs, &acks, id](
+        const std::pair<model::topic, topic_ingestion_policy>& item) {
+          auto& [topic, tip] = item;
+          const errc r = check_topic_policy(topic, tip);
+          if (r != errc::success) {
+              acks.push_back(r);
+              return;
+          }
+          auto logs = _api.local().log_mgr().get(
+            model::topic_namespace(cluster::kafka_namespace, topic));
+          if (logs.empty()) {
+              acks.push_back(errc::topic_does_not_exist);
+              return;
+          }
+          for (auto& [ntp, log] : logs) {
+              auto found = _ntps.find(ntp);
+              if (found != _ntps.end()) {
+                  auto ntp_ctx = found->second;
+                  ntp_ctx->offsets.emplace(id, model::offset(0));
+                  ctxs.emplace(ntp, ntp_ctx);
+              } else {
+                  auto ctx = ss::make_lw_shared<ntp_context>(log, ntp, id);
+                  ctxs.emplace(ntp, ctx);
+                  _ntps.emplace(ntp, ctx);
+              }
+          }
+          acks.push_back(errc::success);
+      });
+}
+
+ss::future<errc> pacemaker::remove_source(script_id id) {
+    return _state_mtx.with([this, id] {
+        auto found = _scripts.find(id);
+        if (found == _scripts.end()) {
+            return ss::make_ready_future<errc>(errc::script_id_does_not_exist);
+        }
+        script_context& ctx = found->second;
+        return ctx.shutdown()
+          .then([this] {
+              /// shutdown explicity clears out strong references to ntp
+              /// contexts. It is known to remove them from the pacemakers cache
+              /// when the use_count() == 1, as there are then known to be no
+              /// more subscribing scripts for the ntp
+              absl::flat_hash_set<model::ntp> erase_set;
+              for (auto& [ntp, ref] : _ntps) {
+                  if (ref.use_count() == 1) {
+                      erase_set.insert(ntp);
+                  }
+              }
+              for (const model::ntp& ntp : erase_set) {
+                  _ntps.erase(ntp);
+              }
+          })
+          .then([this, found] {
+              /// Finally, remove the script_context itself
+              _scripts.erase(found);
+              return errc::success;
+          });
+    });
+}
+
+ss::future<bool> pacemaker::local_script_id_exists(script_id id) {
+    return _state_mtx.with(
```

Comment:
@mmaslankaprv true, good point, however would that also be the case if i make the change to `do_add_source()` to not use a seastar loop and just return void instead of ss::now()? In other words, do you need to use a seastar mutex if portion of code doesn't contain any futures. In this way the reactor couldn't yield to another future during the execution of that method and there wouldn't be any contention.